pkg/clients/exec_command.go

```diff
@@ -155,3 +164,210 @@ func (c *ContainerExecContext) ExecCommand(command []string) (stdout, stderr str
+		context.TODO(),
+		pod,
+		metav1.CreateOptions{},
+	)
+	c.pod = pod
+	if err != nil {
+		return fmt.Errorf("failed to create pod: %w", err)
+	}
+	return nil
+}
+
+func (c *ContainerCreationExecContext) listPods(options *metav1.ListOptions) (*corev1.PodList, error) {
+	pods, err := c.clientset.K8sClient.CoreV1().Pods(c.pod.Namespace).List(
+		context.TODO(),
+		*options,
+	)
+	if err != nil {
+		return pods, fmt.Errorf("failed to find pods: %s", err.Error())
+	}
+	return pods, nil
+}
+
+func (c *ContainerCreationExecContext) refeshPod() error {
+	pods, err := c.listPods(&metav1.ListOptions{
+		FieldSelector:   fields.OneTermEqualSelector("metadata.name", c.podName).String(),
+		ResourceVersion: c.pod.ResourceVersion,
+	})
+	if err != nil {
+		return err
+	}
+	if len(pods.Items) == 0 || len(pods.Items) > 1 {
+		// I don't think k8s allows more than one pod with the same name
+		return errors.New("found multiple pods with the same name")
+	}
+	c.pod = &pods.Items[0]
+
+	return nil
+}
+
+func (c *ContainerCreationExecContext) IsPodRunning() (bool, error) {
+	err := c.refeshPod()
+	if err != nil {
+		return false, err
+	}
+	if c.pod.Status.Phase == corev1.PodRunning {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (c *ContainerCreationExecContext) WaitForPodToStart() error {
+	start := time.Now()
+	for time.Since(start) <= startTimeout {
+		running, err := c.IsPodRunning()
+		if err != nil {
+			return err
+		}
+		if running {
+			return nil
+		}
+	}
+	return errors.New("timed out waiting for pod to start")
+}
+
+func (c *ContainerCreationExecContext) CreatePodAndWaitForStart() error {
+	var err error
+	running := false
+	if c.pod != nil {
+		running, err = c.IsPodRunning()
+		if err != nil {
+			return err
+		}
+	}
+	if !running {
+		err := c.CreatePod()
+		if err != nil {
+			return err
+		}
+	}
+	return c.WaitForPodToStart()
+}
+
+func (c *ContainerCreationExecContext) DeletePod() error {
+	deletePolicy := metav1.DeletePropagationForeground
+	err := c.clientset.K8sClient.CoreV1().Pods(c.pod.Namespace).Delete(
+		context.TODO(),
+		c.pod.Name,
+		metav1.DeleteOptions{
+			PropagationPolicy: &deletePolicy,
+		})
+	if err != nil {
+		return fmt.Errorf("failed to delete pod: %w", err)
+	}
+	return nil
+}
+
+func (c *ContainerCreationExecContext) WaitForPodToDelete() error {
+	start := time.Now()
+	for time.Since(start) <= deletionTimeout {
+		pods, err := c.listPods(&metav1.ListOptions{})
+		if err != nil {
+			return err
+		}
+		found := false
+		for _, pod := range pods.Items { //nolint:gocritic // This isn't my object I can't use a pointer
+			if pod.Name == c.podName {
+				found = true
+			}
+		}
+		if !found {
+			return nil
+		}
+	}
+	return errors.New("pod has not terminated within the timeout")
+}
+
+func (c *ContainerCreationExecContext) DeletePodAndWait() error {
+	err := c.DeletePod()
+	if err != nil {
+		return err
+	}
+	return c.WaitForPodToDelete()
+}
+
+func NewContainerCreationExecContext(
```

Comment:
I think only `NewContainerCreationExecContext`, `DeletePodAndWait` and `CreatePodAndWaitForStart` need to be public, the rest of these methods can be private at this point.

Also rename `DeletePodAndWait` and `CreatePodAndWaitForStart` so they match e.g. `DeletePodAndWait` and `CreatePodAndWait`